File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    MODULAR = 1000000007
    cache = [[0 for i in range(m)] for j in range(n)]
    
    for i, j in puddles:
        cache[j - 1][i - 1] = -1
    
    for i in range(n):
        if cache[i][0] == -1:
            for j in range(i, n):
                cache[j][0] = -1
            break
        else:
            cache[i][0] = 1
            
    for i in range(m):
        if cache[0][i] == -1:
            for j in range(i, m):
                cache[0][j] = -1
            break
        else:
            cache[0][i] = 1
            

    for i in range(1, n):
        for j in range(1, m):
            
            if cache[i][j] == -1:
                continue
                
            if cache[i - 1][j] == -1 and cache[i][j - 1] == -1:
                cache[i][j] = -1
            elif cache[i - 1][j] == -1:
                cache[i][j] = cache[i][j - 1]
            elif cache[i][j - 1] == -1:
                cache[i][j] = cache[i - 1][j]
            else:
                cache[i][j] = (cache[i - 1][j] + cache[i][j - 1]) % MODULAR

    return cache[n - 1][m - 1] if cache[n - 1][m - 1] > 0 else 0

File: 프로그래머스/3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int m, int n, vector<vector<int>> puddles) {
    const int MODULAR = 1000000007;
    vector<vector<int>> cache (n, vector<int>(m, 0));
    
    for (auto coor: puddles) {
        cache[coor[1] - 1][coor[0] - 1] = -1;
    }
    
    for (int i = 0; i < n; i++) {
        if (cache[i][0] == -1) {
            for (int j = i; j < n; j++) {
                cache[j][0] = -1;
            }
            break;
        } else {
            cache[i][0] = 1;
        }
    }
    
    
    for (int i = 0; i < m; i++) {
        if (cache[0][i] == -1) {
            for (int j = i; j < m; j++) {
                cache[0][j] = -1;
            }
            break;
        } else {
            cache[0][i] = 1;
        }
    }
    
    for (int i = 1; i < n; i++) {
        for (int j = 1; j < m; j++) {
            if (cache[i][j] == -1)
                continue;
            
            if (cache[i - 1][j] == -1 && cache[i][j - 1] == -1) {
                cache[i][j] = -1;
            } else if (cache[i - 1][j] == -1) {
                cache[i][j] = cache[i][j - 1];
            } else if (cache[i][j - 1] == -1) {
                cache[i][j] = cache[i - 1][j];
            } else {
                cache[i][j] = (cache[i - 1][j] + cache[i][j - 1]) % MODULAR;
            }
        }
    }
    return cache[n - 1][m - 1] == -1 ? 0:cache[n - 1][m - 1];
}